fix(Core/Event): Rework STV Fishing Extravaganza

schell244 opened the pull request:
## Changes Proposed:
Rework STV Fishing Extravaganza.

## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9532

## SOURCE:
- Heavily inspired by VMangos 
- https://github.com/vmangos/core/blob/9d0aeb7d63b2989686f028b0c5c039b671dc9b25/src/scripts/world/npcs_special.cpp#L1434
- https://wowpedia.fandom.com/wiki/Stranglethorn_Fishing_Extravaganza?oldid=2373763
- https://github.com/vmangos/core/commit/4cd5ca80997a18de296129531158ae3cfb76898f
- https://github.com/vmangos/core/commit/4f739079a44423e31f7c2b895fc6f7068cb86e44

## Tests Performed:
- tested in game

## How to Test the Changes:
- Go to booty bay 
- `.event start 62` (spawns npc's - default time: 13:00)
- `.event start 90` (starts turn in quests, Riggle Bassbait should announce that the event has started - 14:00)
- `.event start 15` (pools are spawned - default time: 14:00)
- To see and accept quests, fishing skill >=150 is required
- Check Riggle Bassbait gossip and Jang before master angler
- Turn in master angler quest -> afterwards Jang quest becomes active, Riggle Bassbait has different gossip
- `.event stop 15` Riggle Bassbait should announce that pools are gone (default time: 16:00)
- `.event stop 90` Quests should become unavailable (default time: 17:00)
- `.event stop 62` All npc's despwan (default time: 17:00)
-  Restart server
- `.event start 62`
- `.event start 90`
- Riggle Bassbait master angler quest should remain disabled
- Riggle Bassbait gossip should remain "we already have winner"
- Event start announcement does not happen again
- After one day, all should be reset.

For manually resetting the event states:
1. Delete variables in table `worldstates` in character database for immediate retesting
```sql
DELETE FROM `worldstates` WHERE  `entry` IN (197,198,199,200);
```

## Known Issues and TODO List:
- World state ids ok?
- Noticed two issues (unrelated to this PR):

1. npc quest flag is not immediately updated, when condition changes. 
   This affects Fishbot 5000 and Jang, when event 90 starts. 
   Nearby players sometimes do not immediately see the quest giver indicator, but the quest are shown in the gossip menu and 
   can be accepted. After relog, quest giver indicator is correctly displayed.  (I only noticed this after converting the scripts to db
   with conditions, cpp script did always immediately show the quest giver indicator).

2. Event time is off by one hour.
    Example: start time should be 13:00 (set in db) but when I use `.event info 62` it shows that next active time is 14:00.
    Could be something similar as it was the case for hourly bells: 
    _[update] added tzset(); for localtime_r() in order to fix timezone/daylight saving time_
    https://github.com/azerothcore/azerothcore-wotlk/pull/9151

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

schell244 commented:
Did some more corrections today, now it's ready to be tested.

WORISX commented:
Looks like it's actually not working, fishbot is working fine, but when event 90 is active and you have 150 skill on fishing there is no announcement and there is no quests available on riggle

schell244 commented:
Thanks for testing. Did you turn in the quest master angler once Riggle before or was the quest not available at all?

WORISX commented:
> Thanks for testing. Did you turn in the quest master angler on Riggle once before or was the quest not available at all? Also have you checked the event in real time, or did you test using `.event start` commands?

No problem man hehe, i tried on both ways, in real time and with commands, in both cases Riggle doesn't announce the start of the event at 2PM and havent' been able to take master angler quest at any case, used in the tests a fresh new created character.

WORISX commented:
I debugged it, the problem is that scheduler reset is causing to not pass any action to UpdateAI, i guess

schell244 commented:
Thank you @WORISX ! I haven't checked in real time, I was only using commands for event start/stop. 
I will have a look later what happens in real time. When using only `.event start/stop` commands, was it working for as expected?

schell244 commented:
Oh nvm, just saw you wrote in both cases. I will re-check it as soon as possible.

schell244 commented:
Just did a clean build, reapplied world (+ sql patch of this PR) and characters db and tested the event with `.event start/stop` commands. For me it looks like this: https://youtu.be/obXCj9CR0U8
Haven't done a realtime test yet.
@WORISX you had different results using the commands?

Maelthyrr commented:
I confirm an issue with the realtime launch of this event. For me event doesn't start. You can try with this; https://www.nirsoft.net/utils/run_as_date.html  It's allow the setup a false date/hours for a program ( use on worldserver, tick box 'move the time forward according to the real time')

schell244 commented:
Thanks both @Maelthyrr and @WORISX for your help! Realtime launch should be fixed now.

schell244 commented:
Just added some more changes: I've noticed that there was already a condition for world state variables implemented:
- removed check at conditions which prevented 0 being a valid value for a world state
- converted jang gossip and quest to db

Kitzunu commented on data/sql/updates/pending_db_world/rev_1658570956441891600.sql:
```diff
@@ -0,0 +1,37 @@
+UPDATE `creature_text` SET `BroadcastTextId` = 10610, `Text`='We have a winner! $n has won FIRST PLACE in the tournament!' WHERE `CreatureID` = 15077 AND `GroupID` = 2;
+
+-- Fishbot 5000 quests: only available when event 90 (Turn-ins) is active
+INSERT INTO `conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry`, `SourceId`, `ElseGroup`, `ConditionTypeOrReference`, `ConditionTarget`, `ConditionValue1`, `ConditionValue2`, `ConditionValue3`, `NegativeCondition`, `ErrorType`, `ErrorTextId`, `ScriptName`, `Comment`) VALUES
+(19,0,8225,0,0,12,0,90,0,0,0,0,0,'','Fishing Extravaganza - Turn-ins active'),
+(19,0,8224,0,0,12,0,90,0,0,0,0,0,'','Fishing Extravaganza - Turn-ins active'),
+(19,0,8221,0,0,12,0,90,0,0,0,0,0,'','Fishing Extravaganza - Turn-ins active');
+
+-- Jang quest: event 90 (Turn-ins) is active and event has a winner (world state id 198)
+INSERT INTO `conditions` (`SourceTypeOrReferenceId`, `SourceGroup`, `SourceEntry`, `SourceId`, `ElseGroup`, `ConditionTypeOrReference`, `ConditionTarget`, `ConditionValue1`, `ConditionValue2`, `ConditionValue3`, `NegativeCondition`, `ErrorType`, `ErrorTextId`, `ScriptName`, `Comment`) VALUES
```
missing deletes here =)

Kitzunu commented on src/server/game/Conditions/ConditionMgr.cpp:
```diff
@@ -2302,11 +2302,6 @@ bool ConditionMgr::isConditionTypeValid(Condition* cond)
         break;
     case CONDITION_WORLD_STATE:
     {
-        if (!sWorld->getWorldState(cond->ConditionValue1))
-        {
-            LOG_ERROR("sql.sql", "World state condition has non existing world state in value1 ({}), skipped", cond->ConditionValue1);
-            return false;
-        }
```
why

schell244 replied:
This check does prevent 0 being used as valid world state value. Working with 0/1 values is handy, since it can be used directly with if like `if (sWorld->getWorldState(STV_FISHING_ANNOUNCE_POOLS_DESPAN)) {...}`
Also this is the first usage of world states at condition db, so I guess allowing 0 should not affect other places.

Kitzunu replied:
Cool

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -207,19 +207,30 @@ class npc_elder_clearwater : public CreatureScript
+/*
+ * Stranglethorn Vale Fishing Extravaganza World States
+ */
+enum fishingExtravaganzaWorldStates
```
```suggestion
enum FishingExtravaganzaWorldStates
```

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -207,19 +207,30 @@ class npc_elder_clearwater : public CreatureScript
+    STV_FISHING_ANNOUNCE_POOLS_DESPAN   = 200
+};
 
+enum riggleBassbait
```
```suggestion
enum RiggleBassbait
```

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -207,19 +207,30 @@ class npc_elder_clearwater : public CreatureScript
     QUEST_MASTER_ANGLER                 = 8193,
 
-    DATA_ANGLER_FINISHED                = 1,
+    GAME_EVENT_FISHING_TURN_INS         = 90,
+    GAME_EVENT_FISHING_POOLS            = 15,
```
```suggestion
    EVENT_FISHING_TURN_INS         = 90,
    EVENT_FISHING_POOLS            = 15,
```

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -207,19 +207,30 @@ class npc_elder_clearwater : public CreatureScript
+    GAME_EVENT_FISHING_POOLS            = 15,
 
-    GAME_EVENT_FISHING                  = 62
+    GOSSIP_TEXT_EVENT_ACTIVE            = 7614,
+    GOSSIP_TEXT_EVENT_OVER              = 7714
```
```suggestion
    GOSSIP_EVENT_ACTIVE            = 7614,
    GOSSIP_EVENT_OVER              = 7714
```

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
-            finishWarning = false;
+            m_uiTimer = 0;
+            auto prevWinTime = sWorld->getWorldState(STV_FISHING_PREV_WIN_TIME);
+            if (time(nullptr) - prevWinTime > DAY)
```
Use GameTime instead of time()

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
-                sWorld->setWorldState(GAME_EVENT_FISHING, 1);
+                if (!me->IsQuestGiver())
+                {
+                    me->SetFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
```
Use SetNpcFlag()

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
+            {
+                if (me->IsQuestGiver())
+                {
+                    me->RemoveFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
```
Use RemoveNpcFlag()

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
+                }
+                if (sWorld->getWorldState(STV_FISHING_ANNOUNCE_EVENT_BEGIN))
+                {
+                    sCreatureTextMgr->SendChat(me, RIGGLE_SAY_START, 0, CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, TEXT_RANGE_ZONE);
```
Why not Say()?

schell244 replied:
This should be a map wide announcement (at least from what I saw on VMangos).
I haven't checked whether the broadcast rage is currently correct (I left this code as it was before), maybe `TEXT_RANGE_ZONE` can be replaced with smth like `TEXT_RANGE_MAP`

Kitzunu replied:
This can still be done with Say() as you determine the range in creature_text.TextRange

Kitzunu replied:
https://www.azerothcore.org/wiki/creature_text#textrange

Kitzunu replied:
Well I guess Yell() would be the proper function

schell244 replied:
I had to use Talk() instead of Yell()/Say() since both functions did expect a string view instead of a text id.

schell244 replied:
Nvm, yell works. Just when passing LANG_UNIVERSAL beside the text id there was no function found.

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
+            {
+                if (sWorld->getWorldState(STV_FISHING_ANNOUNCE_POOLS_DESPAN))
+                {
+                    sCreatureTextMgr->SendChat(me, RIGGLE_SAY_POOLS_END, 0, CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, TEXT_RANGE_ZONE);
```
Say()?

schell244 replied:
same as above, map wide announcement

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
+        if (quest->GetQuestId() == QUEST_MASTER_ANGLER)
         {
-            creature->AI()->DoAction(DATA_ANGLER_FINISHED);
+            creature->RemoveFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
```
RemoveNpcFlag()

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
-            creature->AI()->DoAction(DATA_ANGLER_FINISHED);
+            creature->RemoveFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_QUESTGIVER);
             sCreatureTextMgr->SendChat(creature, RIGGLE_SAY_WINNER, player, CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, TEXT_RANGE_ZONE);
+            sWorld->setWorldState(STV_FISHING_PREV_WIN_TIME, time(nullptr));
```
GameTime

Maelthyrr commented:
Hi, I tested again. Everythin work except one minor issue:

- [ ] I have no announce or yell when event 62 start

Otherwise:
- [X] Events start automatically et properly
- [X] Quests are only available when event 90 is start and when you have at least 150 skill of Fishing
- [X] Winner announcement work
- [X] Except issues mention with quest flag, no other issues find

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
+                }
+                if (sWorld->getWorldState(STV_FISHING_ANNOUNCE_EVENT_BEGIN))
+                {
+                    sCreatureTextMgr->SendChat(me, RIGGLE_SAY_START, 0, CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, TEXT_RANGE_ZONE);
```
```suggestion
                    me->Yell(RIGGLE_SAY_START, LANG_UNIVERSAL);
```

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
+            {
+                if (sWorld->getWorldState(STV_FISHING_ANNOUNCE_POOLS_DESPAN))
+                {
+                    sCreatureTextMgr->SendChat(me, RIGGLE_SAY_POOLS_END, 0, CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, TEXT_RANGE_ZONE);
```
```suggestion
                    me->Yell(CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL);
```

Kitzunu commented on src/server/scripts/World/npcs_special.cpp:
```diff
@@ -231,84 +242,95 @@ class npc_riggle_bassbait : public CreatureScript
         {
-            creature->AI()->DoAction(DATA_ANGLER_FINISHED);
+            creature->RemoveNpcFlag(UNIT_NPC_FLAG_QUESTGIVER);
             sCreatureTextMgr->SendChat(creature, RIGGLE_SAY_WINNER, player, CHAT_MSG_MONSTER_YELL, LANG_UNIVERSAL, TEXT_RANGE_ZONE);
```
```suggestion
            me->Yell(RIGGLE_SAY_WINNER, LANG_UNIVERSAL, player);
```

Kitzunu reviewed (changes requested):
```sql
UPDATE `creature_text` SET `TextRange` = 2 WHERE `CreatureID` = 15077 AND `GroupID` IN (0,1,2);
```
To fix zone wide announcement + my previous suggestions

schell244 commented:
Thanks for review @Kitzunu ❤️

schell244 commented:
> Hi, tested again. Everything work except one minor issue:
> 
>     * [ ]  I have no announce or yell when event 62 start
> 
> 
> Otherwise:
> 
>     * [x]  Events start automatically et properly
> 
>     * [x]  Quests are only available when event 90 is start and when you have at least 150 skill of Fishing
> 
>     * [x]  Winner announcement work
> 
>     * [x]  Except issues mention with quest flag, no other issues find

Thanks @Maelthyrr for re-testing. It's expected that there is no announcement when event 62 starts. Event 62 just spawns the NPC's. The event start announcement happens one hour later when event 90 starts.

acidmanifesto commented:
@Kitzunu we have a pending review request for you. Yes or no to merge?
